crates/ruff_python_formatter/src/expression/expr_bin_op.rs

```diff
@@ -290,3 +272,490 @@ impl NeedsParentheses for ExprBinOp {
+
+    fn deref(&self) -> &Self::Target {
+        FlatBinaryExpressionSlice::from_slice(&self.0)
+    }
+}
+
+/// Binary chain represented as a flat vector where operands are stored at even indices and operators
+/// add odd indices.
+///
+/// ```python
+/// a + 5 * 3 + 2
+/// ```
+///
+/// Gets parsed as:
+///
+/// ```text
+/// graph
+/// +
+/// ├──a
+/// ├──*
+/// │   ├──b
+/// │   └──c
+/// └──d
+/// ```
+///
+/// The slice representation of the above is closer to what you have in source. It's a simple sequence of operands and operators,
+/// entirely ignoring operator precedence (doesn't flatten parenthesized expressions):
+///
+/// ```text
+/// -----------------------------
+/// | a | + | 5 | * | 3 | + | 2 |
+/// -----------------------------
+/// ```
+///
+/// The advantage of a flat structure are:
+/// * It becomes possible to adjust the operator / operand precedence. E.g splitting implicit concatenated strings before `+` operations.
+/// * It allows arbitrary slicing of binary expressions for as long as a slice always starts and ends with an operand.
+///
+/// A slice is guaranteed to always start and end with an operand. The smallest valid slice is a slice containing a single operand.
+/// Operands in multi-operand slices are separated by operators.
+#[repr(transparent)]
+struct FlatBinaryExpressionSlice<'a>([OperandOrOperator<'a>]);
+
+impl<'a> FlatBinaryExpressionSlice<'a> {
+    fn from_slice<'slice>(slice: &'slice [OperandOrOperator<'a>]) -> &'slice Self {
+        debug_assert!(
+            !slice.is_empty(),
+            "Operand slice must contain at least one operand"
+        );
+
+        #[allow(unsafe_code)]
+        unsafe {
+            // SAFETY: `BinaryChainSlice` has the same layout as a slice because it uses `repr(transparent)`
+            &*(slice as *const [OperandOrOperator<'a>] as *const FlatBinaryExpressionSlice<'a>)
+        }
+    }
+
+    fn operators(&self) -> impl Iterator<Item = (OperatorIndex, &Operator<'a>)> {
+        self.0.iter().enumerate().filter_map(|(index, part)| {
+            if let OperandOrOperator::Operator(operator) = part {
+                Some((OperatorIndex::new(index), operator))
+            } else {
+                None
+            }
+        })
+    }
+
+    fn operands(&self) -> impl Iterator<Item = (OperandIndex, &Operand<'a>)> {
+        self.0.iter().enumerate().filter_map(|(index, part)| {
+            if let OperandOrOperator::Operand(operand) = part {
+                Some((OperandIndex::new(index), operand))
+            } else {
+                None
+            }
+        })
+    }
+
+    /// Creates a subslice that contains the operands coming after `last_operator` and up to, but not including the `end` operator.
+    fn between_operators(&self, last_operator: Option<OperatorIndex>, end: OperatorIndex) -> &Self {
+        let start = last_operator.map_or(0usize, |operator| operator.right_operand().0);
+        Self::from_slice(&self.0[start..end.0.get()])
```

Comment:
Should `end.0.get()` be exposed as a method instead of accessing the wrapped type?